### ISSUE TITLE
Can't install paradox: forward-sexp: Scan error: "Containing expression ends prematurely", 1999, 2000

### ISSUE THREAD

**Issue by alphapapa** [Bruce-Connor/paradox#49]
```
Saving file /home/me/.emacs.d/elpa/spinner-1.0/spinner-autoloads.el...
Wrote /home/me/.emacs.d/elpa/spinner-1.0/spinner-autoloads.el
Contacting host: elpa.gnu.org:80
Showing all blocks ... done
Generating autoloads for spinner.el...
forward-sexp: Scan error: "Containing expression ends prematurely", 1999, 2000
```

I upgraded to Emacs 24.4 and am trying to upgrade Paradox.  It's failing.  I don't know why.  :(

**Comment by Bruce-Connor** [comment]
That's quite odd. It seems you're getting an error while installing `spinner.el` (which is a new dependency).

Could you try the following?

1. Go into your `/home/me/.emacs.d/elpa/` directory and delete the `spinner-1.0` directory.
2. Start emacs with `emacs -Q`.
3. Call `list-packages`.
4. Install the `spinner` package.

**Comment by alphapapa** [comment]
I followed those steps and it seems to have installed correctly.  Before I posted here, I tried several times deleting the `spinner-1.0` directory, restarting Emacs, and installing the package again, but it failed as described every time.

Just now, I closed Emacs and restarted it normally, without `-Q`, and installed the `paradox` package.  I got this output at the end of the `Compile-Log`, after warnings about references to free variables:

```
In end of data:
paradox-execute.el:356:1:Warning: the following functions are not known to be
    defined: let-alist, spinner-start

In end of data:
paradox-menu.el:740:1:Warning: the function `paradox-sort-by-package' is not
    known to be defined.
```
I'm guessing those are not good signs, haha.  And indeed, when I try to run `paradox-list-packages`, I get this (not sure where to begin the log, so this may have some extra lines at the start):
```
Wrote /home/me/.emacs.d/elpa/paradox-20150310.1701/paradox-autoloads.el
(No changes need to be saved)
Loading async-bytecomp...done
Loading async...done
Started compiling asynchronously directory /home/me/.emacs.d/elpa/paradox-20150310.1701...
byte-code: End of buffer
byte-code: Beginning of buffer
Overriding package-menu--print-info with paradox--print-info
Overriding package-menu--generate with paradox--generate-menu
Overriding package-menu-mode with paradox-menu-mode
Overriding truncate-string-to-width with paradox--truncate-string-to-width
Contacting host: raw.githubusercontent.com:443
[Paradox] Error downloading the list of repositories. This might be a proxy
paradox--refresh-star-count: Invalid read syntax: "]"
```
Thanks for your help.

**Comment by alphapapa** [comment]
I seem to have worked around the problem by copying the uncorrupted `let-alist` package into my regular `elpa` directory.  But I still have no explanation for why Emacs installed a corrupted `let-alist.el` file in the first place.

**Comment by Bruce-Connor** [comment]
When you look at the two files, do you see any obvious difference between them?

**Comment by alphapapa** [comment]
Yes, I'm sorry, I guess I wasn't clear: the corrupted one is actually a binary file with little-to-no discernable text.  The `file` utility says that it's actually gzip data.  It's almost as if Emacs didn't decompress that file.

Since I worked around it, I've been doing other things, and in the process I needed to reinstall magit.  When I tried, it would always go into an infinite loop while "generating autoloads" for it.  But when I ran `emacs -Q`, then it installed correctly.  And then after restarting Emacs without `-Q`, magit works fine.

I don't know what's going on.  I never had any of these problems before I upgraded to 24.4.  I guess this isn't related to Paradox itself, so I understand if you don't want to mess with this.

**Comment by Bruce-Connor** [comment]
Yes, it looks like you have some deep issue with package installation. I strongly recommend figuring this out, and it's almost certainly something in your config.

**Comment by Bruce-Connor** [comment]
I'll close the issue since I think it's not paradox related. But feel free to post more here, I'm happy to help.

**Comment by alphapapa** [comment]
Thanks.  Well, I wish I knew where to begin, other than spending a long time recreating my config from scratch, testing it line-by-line as I go.  Everything was working fine before I upgraded to 24.4.  :(

**Comment by Bruce-Connor** [comment]
You can start by getting rid of all advices. Then you can try commenting out anything that seems related to the `url` or the `package.el` packages.

**Comment by alphapapa** [comment]
The only advice I have is this, which shouldn't be a problem:
```
;; Fix annoying here-doc autoinserting in shell-script-mode
;; http://emacs.stackexchange.com/q/5336
(defadvice sh--maybe-here-document (around be-smart-about-it activate)
  "Do normal here doc auto insert, but if you type another chevron, revert and leave just <<<."
  (if (and (= (current-column) 1)
	   (looking-back "^<")
	   (looking-at "\nEOF")
	   (save-excursion
	     (forward-line -1)
	     (end-of-line 1)
	     (looking-back "<<EOF")))
      (progn (delete-region (search-backward "EOF") (search-forward "EOF" nil t 2))
	     (insert "<"))
    ad-do-it))
```
I have nothing in there related to the `url` package, and the only lines relating to `package` are:
```
(require 'package)
(setq package-archives '(("gnu" . "http://elpa.gnu.org/packages/")
			 ("marmalade" . "http://marmalade-repo.org/packages/")
			 ("melpa" . "http://melpa.milkbox.net/packages/")))
(package-initialize)
```
Thank you for your help.

**Comment by Bruce-Connor** [comment]
Another thing you can try is setting `debug-on-error` to t. When you get another package installation error, you be given a backtrace about it.

**Comment by alphapapa** [comment]
Ok, back to this issue.  I actually forgot about it until I tried to upgrade some packages again.

To make a long story short, after much hair-pulling and going in circles, I finally figured out that it was being caused by `url-http.el` in my `load-path`.  I don't know why I had that file there, or what it was for, but I guess it was outdated and being used instead of a newer version, and it was causing the `spinner.el` file to not be decompressed, which caused Emacs to fail to parse it, which caused it to fail to install paradox.  When I removed `url-http.el`, I was able to upgrade paradox.

Thanks very much for your help.
